Allow installing parallel linkerd versions and custom install dirs.

memory opened the pull request:
- Add an `INSTALLROOT` var to the install script, defaulting to `${HOME}/.linkerd2`

- Save the downloaded binary into `${INSTALLROOT}/bin/linkerd-${LINKERD2_VERSION}`

- If the same version is already present and the checksum passes, do not redownload

- Symlink `${INSTALLROOT}/bin/linkerd` to the versioned binary

- Fail the script if the curl commands fail

- Update documentation to note the presence of versioned binaries

grampelberg commented:
Maybe even simpler, why not just have all the binaries in `~/.linkerd/bin` with their versions and a symlink, like:

```
.linkerd2
├ bin
│   └── linkerd -> edge-19.8.7/linkerd # symlink to latest
│   └── linkerd-edge-19.8.7
│   └── linkerd-stable-2.4.2
```

This way everything's still on your path so you can either be specific about the version or generic.

alpeb commented:
@memory it seems you misspelled `LINKERD2_VERSION` in the `install` and `install-edge` scripts, in lines 41, 42 and 58

alpeb commented:
I like @grampelberg proposal. I can see for example python binaries being handled like that under `/usr/bin`. Packages with more than one binary (not our case) like java are fitter to the directory-per-version approach.

memory commented:
All reasonable requests.  I've taken a second pass at it - take a look now?

grampelberg commented on run.linkerd.io/public/install:
```diff
@@ -32,19 +33,21 @@ url="https://github.com/linkerd/linkerd2/releases/download/${LINKERD2_VERSION}/$
-  mv "${tmp}/${filename}" ".linkerd2/bin/linkerd"
-  chmod +x ".linkerd2/bin/linkerd"
+  mkdir -p ${INSTALLROOT}/bin
+  cp "${tmp}/${filename}" ${INSTALLROOT}/bin/linkerd-${LINKERD2_VERSION}
```
Shouldn't this be a `mv`?

memory replied:
Oops, quite correct.  Fixed!

grampelberg commented:
@memory having a tough time:

```bash
/ # LINKERD2_VERSION=stable-2.5.0 sh install
Downloading linkerd2-cli-stable-2.5.0-linux...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   622    0   622    0     0   1767      0 --:--:-- --:--:-- --:--:--  1767
100 34.4M  100 34.4M    0     0   734k      0  0:00:48  0:00:48 --:--:--  787k

Download complete!, validating checksum...
Checksum valid.

ln: /root/.linkerd2/bin/linkerd-stable-2.5.0: File exists
```

Also, I just merged another install script update that'll require a rebase to get the merge to work. Sorry this has taken me so long! I was having a tough time figuring out how to test it ... and this morning I just ran `docker run` .... embarrassing.

memory commented:
Oh, sheesh.  Later today or tomorrow I'll try to rebase and make sure it, you know, actually works.

memory commented:
Okay, (a) derp, and (b) I just realized we could save a bit of bandwidth here, so behold:

```
$ LINKERD2_VERSION=stable-2.5.0 run.linkerd.io/public/install
Downloading linkerd2-cli-stable-2.5.0-darwin...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   623    0   623    0     0   5142      0 --:--:-- --:--:-- --:--:--  5106
100 38.5M  100 38.5M    0     0  2360k      0  0:00:16  0:00:16 --:--:-- 2812k
Download complete!

Validating checksum...
Checksum valid.

Linkerd stable-2.5.0 was successfully installed 🎉


Add the linkerd CLI to your path with:

  export PATH=$PATH:/Users/n/.linkerd2/bin

Now run:

  linkerd check --pre                     # validate that Linkerd can be installed
  linkerd install | kubectl apply -f -    # install the control plane into the 'linkerd' namespace
  linkerd check                           # validate everything worked!
  linkerd dashboard                       # launch the dashboard

Looking for more? Visit https://linkerd.io/2/next-steps
```

But if we run it again: 

```
$ LINKERD2_VERSION=stable-2.5.0 run.linkerd.io/public/install

Validating checksum...
Checksum valid.

Linkerd stable-2.5.0 was already downloaded; making it the default 🎉

To force re-downloading, delete '/Users/n/.linkerd2/bin/linkerd-stable-2.5.0' then run me again.

Add the linkerd CLI to your path with:

  export PATH=$PATH:/Users/n/.linkerd2/bin

Now run:

  linkerd check --pre                     # validate that Linkerd can be installed
  linkerd install | kubectl apply -f -    # install the control plane into the 'linkerd' namespace
  linkerd check                           # validate everything worked!
  linkerd dashboard                       # launch the dashboard

Looking for more? Visit https://linkerd.io/2/next-steps
```

...but if we corrupt the binary:

```
$ echo x >> ~/.linkerd2/bin/linkerd-stable-2.5.0;  LINKERD2_VERSION=stable-2.5.0 run.linkerd.io/public/install

Validating checksum...
Checksum validation failed.
Downloading linkerd2-cli-stable-2.5.0-darwin...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   623    0   623    0     0   4247      0 --:--:-- --:--:-- --:--:--  4267
100 38.5M  100 38.5M    0     0  2342k      0  0:00:16  0:00:16 --:--:-- 3209k
Download complete!

Validating checksum...
Checksum valid.

Linkerd stable-2.5.0 was successfully installed 🎉


Add the linkerd CLI to your path with:

  export PATH=$PATH:/Users/n/.linkerd2/bin

Now run:

  linkerd check --pre                     # validate that Linkerd can be installed
  linkerd install | kubectl apply -f -    # install the control plane into the 'linkerd' namespace
  linkerd check                           # validate everything worked!
  linkerd dashboard                       # launch the dashboard

Looking for more? Visit https://linkerd.io/2/next-steps
```

Also I cleaned up a bunch of hard-to-read variables and added the `-f` flag to the curl commands so that if those fail the script won't break with daft checksum verification errors.

memory commented:
(The circleci failure appears to be a problem with an upstream dependency that is not related to this PR)
